drv/stm32h7-usart/src/lib.rs

```diff
@@ -0,0 +1,139 @@
+
+#[cfg(feature = "h753")]
+pub use stm32h7::stm32h753 as device;
+
+use drv_stm32xx_sys_api::{Alternate, Peripheral, PinSet, Sys};
+use unwrap_lite::UnwrapLite;
+
+/// Handle to an enabled USART device.
+pub struct Usart {
+    usart: &'static device::usart1::RegisterBlock,
+}
+
+impl Usart {
+    /// Turn on the `USART` described by `usart`, `peripheral`, `tx_rx_mask`,
+    /// and `alternate`, with the baud rate defined by `clock_hz` and
+    /// `baud_rate`.
+    ///
+    /// Enables interrupts from the `USART` when bytes are available to receive;
+    /// the caller is responsible for enabling and handling the corresponding
+    /// kernel interrupt.
+    pub fn turn_on(
+        sys: &Sys,
+        usart: &'static device::usart1::RegisterBlock,
+        peripheral: Peripheral,
+        tx_rx_mask: PinSet,
+        alternate: Alternate,
+        clock_hz: u32,
+        baud_rate: u32,
+    ) -> Self {
+        // Turn the actual peripheral on so that we can interact with it.
+        sys.enable_clock(peripheral);
+        sys.leave_reset(peripheral);
+
+        // Set FIFO interrupt thresholds; must be done before enabling the
+        // UART.
+        //
+        // Two TODOs/questions here:
+        // 1. Do we want the caller to be able to choose the threshold?
+        // 2. Do we want a threshold on the RX fifo? for now, use the
+        //    `rxne()` bit to check for nonempty (instead of waiting for the
+        //    FIFO to reach a threshold)
+        //
+        // Safety: 0b11x values are undefined; see RM0433 48.7.4. We're using a
+        // fixed, defined value.
+        unsafe {
+            // 0b101 == interrupt when TX fifo is completely empty
+            usart.cr3.write(|w| w.txftcfg().bits(0b101));
+        }
+
+        // Enable the UART in FIFO mode.
+        usart.cr1.write(|w| w.fifoen().set_bit().ue().enabled());
+
+        // set the baud rate
+        let cycles_per_bit = (clock_hz + (baud_rate / 2)) / baud_rate;
+        usart.brr.write(|w| w.brr().bits(cycles_per_bit as u16));
+
+        // Enable the transmitter and receiver.
+        usart.cr1.modify(|_, w| w.te().enabled().re().enabled());
+
+        sys.gpio_configure_alternate(
+            tx_rx_mask,
+            drv_stm32xx_sys_api::OutputType::PushPull,
+            drv_stm32xx_sys_api::Speed::Low,
+            drv_stm32xx_sys_api::Pull::None,
+            alternate,
+        )
+        .unwrap_lite();
+
+        // Enable RX interrupts from the USART side.
+        usart.cr1.modify(|_, w| w.rxneie().enabled());
+
+        Self { usart }
+    }
+
+    /// Try to push `byte` into the USART's TX FIFO, returning `true` on success
+    /// or `false` if the FIFO is currently full.
+    pub fn try_tx_push(&self, byte: u8) -> bool {
+        // Check if TX fifo is not full
+        if self.usart.isr.read().txe().bit() {
+            // Stuff byte into TX fifo.
+            self.usart.tdr.write(|w| w.tdr().bits(u16::from(byte)));
```

Comment:
Why are we wrapping a byte in a `u16` here?

Reply:
Short answer - that's the type required by `w.tdr().bits(_)`!

Longer answer - it looks like the TDR register accepts 9 bits, but the 9th bit is ignored/replaced by the parity calculation if parity calculations are enabled (which they are by default). I assume the function takes a u16 to account for that.